### PR TITLE
P11 kwd class

### PR DESCRIPTION
**Pull request by Alpana07**
This commit implements concrete class to parse keyword VPD.

### CODE REVIEW

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/keyword_vpd_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        validateKwVpdDataEndTag();
+        validateChecksum();
+    }
+    validateKwVpdEndTag();
+
+    return kwValMap;
+}
+
+int KeywordVpd::checkIsBonoVpd()
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:36]: A bool should do over here?

File: src/keyword_vpd_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    // +TWO_BYTES is the description's size byte
+    std::advance(vpdIterator, constants::TWO_BYTES + dataSize);
+
+    int kwVpdType = 0;
+    // Check for invalid vendor defined large resource type
+    if (*vpdIterator != constants::KW_VPD_START_TAG)
+    {
+        if (*vpdIterator != constants::ALT_KW_VPD_START_TAG)
+        {
+            throw std::runtime_error("Invalid Keyword Vpd Start Tag");
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:51]: Why runtime error and not VPD data exception?

File: src/keyword_vpd_parser.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+int KeywordVpd::checkIsBonoVpd()
+{
+    size_t dataSize = getKwDataSize();
+
+    checkNextBytesValidity(constants::TWO_BYTES + dataSize);
+
+    // +TWO_BYTES is the description's size byte
+    std::advance(vpdIterator, constants::TWO_BYTES + dataSize);
+
+    int kwVpdType = 0;
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:45]: Why int needed here?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: include/constants.hpp [hunk 4]
```diff
@@ -5,13 +5,20 @@ namespace vpd
 {
 namespace constants
 {
-static constexpr auto FORMAT_11S_LEN = 3;
-static constexpr auto MEMORY_VPD_DATA_START = 416;
-static constexpr int IPZ_DATA_START = 11;
-static constexpr uint8_t KW_VAL_PAIR_START_TAG = 0x84;
-static constexpr int KW_VPD_DATA_START = 0;
+static constexpr uint8_t IPZ_DATA_START_AT = 11;
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply include/constants.hpp:8]: This is to categorise IPZ, KW, memory constants at one place.
like, all IPZ_* , all KW_* together , is easy readable and looks neat.

File: src/keyword_vpd_parser.hpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "parser_interface.hpp"
+#include "types.hpp"
+
+namespace vpd
+{
+namespace keyword
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:8]: ok

File: src/keyword_vpd_parser.hpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+{
+
+/**
+ * @brief Concrete class to implement Keyword VPD parsing
+ *
+ * The class inherits ParserInterface interface class and overrides the parser
+ * functionality to implement parsing logic for Keyword VPD format.
+ */
+class KeywordVpd : public ParserInterface
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:17]: ok

File: src/keyword_vpd_parser.hpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+     *
+     * Move kwVpdVector to parser object's kwVpdVector
+     */
+    KeywordVpd(const types::BinaryVector& kwVpdVector) :
+        keywordVpdVector(kwVpdVector)
+    {
+    }
+
+    /**
+     * @brief Parse the keyword VPD binary data.
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:36]: ok

File: src/keyword_vpd_parser.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+     * Calls the sub functions to emplace the
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:37]: ok

File: src/keyword_vpd_parser.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+     * keyword-value pairs in map and to validate
+     * certain tags and checksum data.
+     *
+     * @return map of keyword:value
+     */
+    std::variant<types::KeywordVpdMap, types::ParsedData> parse();
+
+    /**
+     * @brief An api to return interface name with respect to
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:46]: yes. corrected.

File: src/keyword_vpd_parser.hpp [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+     * the parser selected.
+     *
+     * @return - Interface name for that vpd type.
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:49]: yes, it was removed from cpp file. will remove from hpp also.

File: src/keyword_vpd_parser.hpp [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    /*Vector of keyword VPD data*/
+    const types::BinaryVector& keywordVpdVector;
+
+
+    /**
+     * @brief Validate the type of keyword VPD
+     *
+     * @return integer representing the type of kw VPD.
+     */
+    int checkIsBonoVpd();
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:73]: sure..done

File: src/keyword_vpd_parser.cpp (hunk 1 above)
Comment (Alpana07) [reply src/keyword_vpd_parser.cpp:36]: yes

File: src/keyword_vpd_parser.hpp [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * @brief Parse the keyword VPD binary data.
+     * Calls the sub functions to emplace the
+     * keyword-value pairs in map and to validate
+     * certain tags and checksum data.
+     *
+     * @return map of keyword:value
+     */
+    std::variant<types::KeywordVpdMap, types::ParsedData> parse();
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply src/keyword_vpd_parser.hpp:43]: done

File: src/keyword_vpd_parser.cpp (hunk 3 above)
Comment (Alpana07) [reply src/keyword_vpd_parser.cpp:45]: done

File: include/constants.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply include/constants.hpp:8]: Please don't add AT atth end.

File: include/keyword_vpd_parser.hpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    types::BinaryVector::const_iterator m_checkSumStart;
+
+    /*Iterator pointing to checksum end byte*/
+    types::BinaryVector::const_iterator m_checkSumEnd;
+
+    /*Iterator to vpd data*/
+    types::BinaryVector::const_iterator m_vpdIterator;
+
+    /*Vector of keyword VPD data*/
+    const types::BinaryVector& m_keywordVpdVector;
```
Comment (SunnySrivastava1984) [inline include/keyword_vpd_parser.hpp:97]: Class member variables should not be left uninitialized once the class has been instantiated.

File: src/keyword_vpd_parser.cpp [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+void KeywordVpdParser::checkNextBytesValidity(uint8_t numberOfBytes)
+{
+
+    if ((std::distance(m_keywordVpdVector.begin(),
+                       m_vpdIterator + numberOfBytes)) >
+        std::distance(m_keywordVpdVector.begin(), m_keywordVpdVector.end()))
+    {
+        throw(DataException("Truncated VPD data"));
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:118]: Handle all the exceptions in the base function of the same class itself. Re-throw the exceptions if required to the main caller.

File: src/keyword_vpd_parser.cpp [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+    checkNextBytesValidity(sizeof(constants::KW_VPD_START_TAG));
+    std::advance(m_vpdIterator, sizeof(constants::KW_VPD_START_TAG));
+
+    size_t dataSize = getKwDataSize();
+
+    checkNextBytesValidity(constants::TWO_BYTES + dataSize);
+
+    // +TWO_BYTES is the size byte
+    std::advance(m_vpdIterator, constants::TWO_BYTES + dataSize);
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:26]: Why these checks were moved to this function ( line 16 to 26) it was there in separate function which was more readable.

File: src/keyword_vpd_parser.cpp [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        throw(DataException("Invalid Small resource type."));
+    }
+
+    return kwValMap;
+}
+
+types::KeywordVpdMap KeywordVpdParser::parser()
+{
+    int totalSize = getKwDataSize();
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:56]: Can this be made more efficient? Why int is required?

File: src/keyword_vpd_parser.cpp [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace vpd
+{
+
+/* @brief Keyword VPD parser function*/
+std::variant<types::ParsedVPD, types::KeywordVpdMap> KeywordVpdParser::parse()
+{
+    m_vpdIterator = m_keywordVpdVector.begin();
+
+    checkNextBytesValidity(sizeof(constants::KW_VPD_START_TAG));
```
Comment (SunnySrivastava1984) [inline src/keyword_vpd_parser.cpp:18]: Function name is ambiguous. "checkNextBytesValidity" for what?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
Please build and test the code for any runtime validation check.

File: include/constants.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alpana07) [reply include/constants.hpp:8]: ok.

File: src/keyword_vpd_parser.cpp (hunk 2 above)
Comment (Alpana07) [reply src/keyword_vpd_parser.cpp]: Ack.

**Pull request closed without merging**